pkg/metrics/composite_tags.go

```diff
@@ -0,0 +1,115 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022-present Datadog, Inc.
+
+package metrics
+
+import (
+	"encoding/json"
+	"strings"
+)
+
+// CompositeTags stores two tag sets and provides methods to manipulate them easily
+type CompositeTags struct {
+	tags1 []string
+	tags2 []string
+}
+
+// NewCompositeTags creates a new CompositeTags
+func NewCompositeTags(tags1 []string, tags2 []string) *CompositeTags {
+	return &CompositeTags{
+		tags1: tags1,
+		tags2: tags2,
+	}
+}
+
+// CompositeTagsFromSlice creates a new CompositeTags from a slice
+func CompositeTagsFromSlice(tags []string) *CompositeTags {
+	return NewCompositeTags(tags, nil)
+}
+
+// CombineCompositeTagsAndSlice creates a new CompositeTags from an existing CompositeTags and string slice
+func CombineCompositeTagsAndSlice(compositeTags *CompositeTags, tags []string) *CompositeTags {
+	if compositeTags.tags2 == nil {
+		return NewCompositeTags(compositeTags.tags1, tags)
+	}
+	// Copy tags in case `CombineCompositeTagsAndSlice` is called twice with the same first argument.
+	// For example see TestCompositeTagsCombineCompositeTagsAndSlice.
+	newTags := append(append([]string{}, compositeTags.tags1...), tags...)
+	return NewCompositeTags(newTags, compositeTags.tags2)
```

Comment:
As I understand it, the idea here is to avoid unnecessary copying of slices, so it's an important part of the API to know _which_ slices will get copied in this operation.  I assume that's part of why tags1 and the argument slice are copied, while tags2 is not?

Reply:
@vickenty Code updated: `tags` is appended to `tags2`

@djmitche : The copy is required for correctness. If I replace  
```
newTags := append(append([]string{}, compositeTags.tags1...), tags...)
``` 
by 
```
newTags := append(compositeTags.tags1..., tags...)
```
 then `TestCompositeTagsCombineCompositeTagsAndSlice` failed.
A simple example of this issue is https://go.dev/play/p/URRs_T7VhR5. 
From my point of view the caller should not be aware of the implementation and shoudn't make assumption about which slice is copied.

Reply:
I agree it's required -- I was suggesting documenting it, since the purpose of this API is performance and copies aren't terribly performant :)

Reply:
Comment added.